students/Nick_Haury/Lesson06/mailroom4.py

```diff
@@ -0,0 +1,317 @@
+                    "4": quit
+    }
+
+    print("Welcome to the mailroom program")
+    prompt = "\n".join((
+        "Please choose a function from the below options:",
+        "1 - Send a Thank You to a single donor.",
+        "2 - Create a Report.",
+        "3 - Write letters to all donors to file.",
+        "4 - Quit",
+        ">> "
+        ))
+
+    while True:
+        choice = input(prompt)
+        try:
+            choice_dict[choice]()
+        except KeyError:
+            print(str(choice) + " is an invalid selection, "
+                  "please make a valid selection \n")
+
+
+def thank_you():
+    '''
+    prompts the user for a full name.  If the user types "list", then print
+    a list of donor names and re-prompt.  If a name is typed that is not
+    in the list, add the name to the list.  If a name is typed that is
+    already in the list, then use it.  Once a name is selected, a
+    donation amount is specified.  Donation amount input by users is then
+    added to the donation history of the name.  Email is then automatically
+    drafted using string formatting thanking the donor for the donation.
+    Return to main prompt after.
+    '''
+
+    print()
+    while True:
+        ty_input = input('Type "list" to display a list of donor names. '
+                         'Otherwise enter a name you would like to add a '
+                         'donation for:\n>>')
+
+        if ty_input.lower() == "list":
+            print("\nCurrent Donors: ")
+            for row in get_donor_list():
+                print(row)
+            print()
+        else:
+            try:
+                donation_amount = float(input("\nPlease enter donation amount:"
+                                              "\n>>"))
+                if donation_amount <= 0:
+                    raise ValueError
+            except ValueError:
+                print("\nPlease only enter a single positive number for "
+                      "donation amount\n")
+            else:
+                add_donation(ty_input, donation_amount)
+                print(
+                    "\n", create_email_text(ty_input, donation_amount), sep=""
+                    )
+                break
+
+
+def get_donor_list(donors=donors):
+    return list(donors.keys())
+
+
+def create_email_text(donor_name, donation_amount):
+    '''
+    Takes a donor's name and donation amount as parameters, and then creates
+    an email template thanking them for their donation
+    '''
+
+    return(f"Dear {donor_name},\n\n"
+           f"It is with incredible gratitude that we accept your wonderfully "
+           f"generous donation of ${donation_amount:,.2f}.  Your "
+           "contribution will truly make a difference in the path forward "
+           "towards funding our common goal."
+           "\n\nEver Greatefully Yours,\n\n"
+           "X" + ("_" * 20) + "\n")
+
+
+def add_donation(donor_name, donation_amount, donors=donors):
```

Comment:
Since I have my donors dictionary as a global, it seemed like the only way I could use custom donor dictionaries in my tests was to include a keyword argument for the dictionary.  I'm not sure if this is good practice or not though... I think I remember one of the readings discussing creating dummies somehow; may need to investigate that further if this is bad practice.

Reply:
yes, in your test file you can override your global data structure

```
import mailroom
mailrom.donors = {}
```